### PR TITLE
Upload in 4mb chunks

### PR DESCRIPTION
**Pull request by johnheroy**
Resolves https://github.com/fractalwagmi/fractal-cli/issues/13

Upload in 4MB chunks (serial) with up to 3 retries.

### CODE REVIEW

File: pkg/storage/upload.go [hunk 1]
```diff
@@ -1,34 +1,61 @@
...
-	if err != nil {
-		return err
+	log.Println("Uploading file...")
+
+	// Upload file in chunks with retry policy for resilience against transient connectivity errors.
+	for offset := int64(0); offset < fileSize; offset += chunkSize {
+		chunk := make([]byte, chunkSize)
+		n, err := file.Read(chunk)
+		if err != nil && err != io.EOF {
+			return err
```
Comment (ricebin) [inline pkg/storage/upload.go:43]: Don't need both log fatal and return err
Cuz log fatal crashes

File: pkg/storage/upload.go [hunk 2]
```diff
@@ -47,7 +74,67 @@ func getResumableUploadUrl(url string) (string, error) {
...
+
+	chunkSize := int64(len(chunk))
+	byteRange := fmt.Sprintf("bytes %d-%d/%d", offset, offset+chunkSize-1, fileSize)
+	lastChunk := offset+defaultChunkSize >= fileSize
+
+	req.Header.Set("Content-Length", fmt.Sprint(chunkSize))
+	req.Header.Set("Content-Range", byteRange)
+
+	res, err := http.DefaultClient.Do(req)
+	if err != nil {
```
Comment (ricebin) [inline pkg/storage/upload.go:127]: Todo
Can we pass thru an instance of http client so we can configure it later

**Review by ricebin — APPROVED** [review]

File: pkg/storage/upload.go [hunk 3]
```diff
@@ -1,34 +1,61 @@
...
 	if err != nil {
 		return err
 	}
-	req.Header.Set("Content-Type", "application/zip")
+	fileSize := fileInfo.Size()
 
-	res, err := http.DefaultClient.Do(req)
-	if err != nil {
-		return err
+	log.Println("Uploading file...")
```
Comment (ricebin) [inline pkg/storage/upload.go:36]: We would use just std println here no?

**Review by ricebin — APPROVED** [review]

File: pkg/storage/upload.go [hunk 4]
```diff
@@ -1,34 +1,61 @@
...
+
+	// Upload file in chunks with retry policy for resilience against transient connectivity errors.
+	for offset := int64(0); offset < fileSize; offset += defaultChunkSize {
+		chunk := make([]byte, defaultChunkSize)
+		n, err := file.Read(chunk)
+		if err != nil && err != io.EOF {
```
Comment (ricebin) [inline pkg/storage/upload.go:42]: Some is fishy here 
What if we still have x bytes and hit eof here?

File: pkg/storage/upload.go (hunk 1 above)
Comment (johnheroy) [reply pkg/storage/upload.go:43]: I got rid of the log.Fatal calls (right after pushing the PR) in this file, what do u think about only returning errs here?

Comment (ricebin) [reply pkg/storage/upload.go:43]: i prefer to only log.Fatal at the outer layer (/cmd)

File: pkg/storage/upload.go (hunk 2 above)
Comment (johnheroy) [reply pkg/storage/upload.go:122]: done - completed in another PR

File: pkg/storage/upload.go (hunk 3 above)
Comment (johnheroy) [reply pkg/storage/upload.go:31]: Yeah we can

File: pkg/storage/upload.go (hunk 4 above)
Comment (johnheroy) [reply pkg/storage/upload.go:36]: we would keep going, and upload that last chunk if we hit EOF. is there a case where we get EOF but there are more bytes that weren't read yet?

**Pull request merged**